src/image_builder.py

```diff
@@ -154,109 +156,198 @@ def image_builder(buildspec):
+            CONCLUSION_STAGE_IMAGES.append(conclusion_stage_image_object)
 
-        # the FORMATTER.progress(THREADS) function call also waits until all threads have completed
-        FORMATTER.progress(THREADS)
+    FORMATTER.banner("DLC")
+    FORMATTER.title("Status")
+    
+    # Standard images must be built before example images
+    # Example images will use standard images as base
+    # Conclusion images must be built at the end as they will consume respective standard and example images
+    standard_images = [image for image in INITIAL_STAGE_IMAGES if "example" not in image.name.lower()]
+    example_images = [image for image in INITIAL_STAGE_IMAGES if "example" in image.name.lower()]
+    conclusion_stage_images = [image for image in CONCLUSION_STAGE_IMAGES]
+    ALL_IMAGES = INITIAL_STAGE_IMAGES + CONCLUSION_STAGE_IMAGES
+    IMAGES_TO_PUSH = [image for image in ALL_IMAGES if image.to_push and image.to_build]
+
+    #initial stage standard images build
+    FORMATTER.banner("Standard Build")
+    build_images(standard_images)
+
+    #initial stage example images build
+    FORMATTER.banner("Example Build")
+    build_images(example_images)
+       
+    #Conclusion stage build
+    if len(conclusion_stage_images) > 0:
+        FORMATTER.banner("Conclusion Build")
+        build_images(conclusion_stage_images, make_dummy_boto_client=True)
+    
+    FORMATTER.banner("Push Started")
+    push_images(IMAGES_TO_PUSH)
+
+    FORMATTER.title("Log Display")
+    #After the build, display logs/summary for all the images.
+    show_build_logs(ALL_IMAGES)
+    show_build_summary(ALL_IMAGES)
+    is_any_build_failed, is_any_build_failed_size_limit = show_build_errors(ALL_IMAGES)
+
+    #From all images, filter the images that were supposed to be built and upload their metrics
+    BUILT_IMAGES = [image for image in ALL_IMAGES if image.to_build]
+
+    FORMATTER.title("Upload Metrics")
+    # change logic here. upload metrics only for the Conclusion stage image
+    upload_metrics(BUILT_IMAGES, BUILDSPEC, is_any_build_failed, is_any_build_failed_size_limit)
+
+    FORMATTER.title("Setting Test Env")
+    # Set environment variables to be consumed by test jobs
+    test_trigger_job = utils.get_codebuild_project_name()
+    # Tests should only run on images that were pushed to the repository
+    utils.set_test_env(
+        IMAGES_TO_PUSH,
+        BUILD_CONTEXT=os.getenv("BUILD_CONTEXT"),
+        TEST_TRIGGER=test_trigger_job,
+    )
+
+def get_conclusion_stage_image_object(initial_stage_image_object):
+    # Check if this is only required for mainline
+    # if build_context == "MAINLINE":
+    conclusion_stage_image_object = None
+    conclusion_stage_image_object = ConclusionStageImage(
+        info=initial_stage_image_object.info,
+        dockerfile=os.path.join(os.sep, os.getenv("ROOT_FOLDER_PATH"), "src", "Dockerfile.multipart"),
+        repository=initial_stage_image_object.repository,
+        tag=initial_stage_image_object.tag,
```

Comment:
optionally we can also have a additional tag that will be applied to the `conclusion` stage image so that we can easily distinguish between a single/mulitstage stage DLC

Reply:
I dont think there is will be an issue since we will be adding a new tag and not modifying an existing tag.

Reply:
Done.

Reply:
Add `-`